### PR TITLE
`RequestHandler.reply`'s parameter data converted to function

### PR DESCRIPTION
**Pull request by wangdengwu**
# Pull Request Template

## Your checklist for this pull request

Please review the [guidelines for contributing](https://github.com/lomsa-dev/http-mock-adapter/blob/develop/CONTRIBUTING.md) to this repository.

- [x] Make sure you are requesting to **pull a `feature`/`bugfix` branch** (right side). Don't request your `main` or `develop`!
- [x] Make sure you are making a pull request against the **`develop` branch** (left side). Also, you should start *your branch* off *our `develop`*.
- [x] Make sure that you add reviewers, assignees, labels, projects, milestones and linkes issues to the pull request as you see fit.
- [x] Check the commit's or even all commits' message styles to match our requested structure.
- [x] Check your code additions will fail neither code linting checks nor unit test.

**Tip**: Delete everything but the following section that contains the description of the pull request.

---

## Description
For mock http request, not just for unit test. so RequestHandler.reply's parameter data should be function.
so when the real RequestOptions pass in ,the return data can be code with logic.

Please describe your pull request.

Thank you!

### CODE REVIEW

File: lib/src/extensions/matches_request.dart [hunk 1]
```diff
@@ -50,6 +51,10 @@ extension MatchesRequest on RequestOptions {
       return true;
     }
 
+    /// if data is MockDataCallback do not need to match;
+    if (expected is MockDataCallback) {
```
Comment (theiskaa) [inline lib/src/extensions/matches_request.dart:55]: We should replace 
```dart
if (expected is MockDataCallback) return true;
```

**Review by theiskaa — APPROVED** [review]
Everything is ok, but we should make some little changes in code to make it better.

**Comment by LukaGiorgadze** [comment]
Thanks @wangdengwu for contributing and helping **http-mock-adapter**.
Looks good, but it seems some CI/CD tools is failing, including linter. Please fix that and we will re-review then.

@theiskaa Btw, have you a chance to test it on your device/simulator?

File: lib/src/extensions/matches_request.dart [hunk 2]
```diff
@@ -50,6 +51,10 @@ extension MatchesRequest on RequestOptions {
...
+      return true;
+    }
```
Comment (theiskaa) [inline lib/src/extensions/matches_request.dart:57]: Sorry, I forgot to mention some little changes to make code better look.

Please change to:
```dart
//  If data is MockDataCallback do not need to match.
if (expected is MockDataCallback) return true
```

File: test/extensions/matches_request_test.dart [hunk 3]
```diff
@@ -36,6 +36,15 @@ void main() {
         );
       });
 
+      test('data is MockDataCallback', () {
+        const actual = {
+          'a': 'a',
+          'b': 'b',
```
Comment (theiskaa) [inline test/extensions/matches_request_test.dart:42]: Remove unneeded comma here. So, the variable should be like this:
```dart
const actual = {'a': 'a', 'b': 'b'}
```

File: test/mixins/recording_test.dart [hunk 4]
```diff
@@ -43,6 +43,27 @@ void main() {
       }
     });
 
+    test('test MockDataCallback', () async {
+      var data = {
+        'id': 1,
+        'name': 'mock',
```
Comment (theiskaa) [inline test/mixins/recording_test.dart:49]: Remove unneeded comma here. So, the variable should be like this:
```dart
  var data = {'id': 1,  'name': 'mock'}
```

File: test/handlers/request_handler_test.dart [hunk 5]
```diff
@@ -45,6 +46,34 @@ void main() {
       );
     });
 
+    test('sets response data MockDataCallback ', () async {
+      const statusCode = HttpStatus.ok;
+      var data = {'data': 'OK'};
+      dynamic Function(RequestOptions options) inputData;
+      inputData = (options) {
+        return data;
+      };
```
Comment (theiskaa) [inline test/handlers/request_handler_test.dart:55]: I don't like this frequency.

File: test/handlers/request_handler_test.dart [hunk 6]
```diff
@@ -45,6 +46,34 @@ void main() {
...
+
+      requestHandler.reply(
+        statusCode,
+        inputData,
```
Comment (theiskaa) [inline test/handlers/request_handler_test.dart:59]: Remove unneeded comma here. So, the request should be like this:
```dart
requestHandler.reply(statusCode, inputData);
```

**Review by theiskaa — CHANGES_REQUESTED** [review]
I put some suggestions to make code look better.

**Comment by theiskaa** [comment]
> Thanks @wangdengwu for contributing and helping **http-mock-adapter**.
> Looks good, but it seems some CI/CD tools is failing, including linter. Please fix that and we will re-review then.
> 
> @theiskaa Btw, have you a chance to test it on your device/simulator?

I'll do it now ... Sorry, I was confused, my fault :/

**Comment by wangdengwu** [comment]
got it 
i will clean the code then pull request again
thanks for review 🙂

**Review by theiskaa — APPROVED** [review]
Looks good. 

@GiorgiBeriashvili could you look main idea of PR and tell your thoughts?
Code is ok, just we need review for main idea.

@wangdengwu Thanks for PR 🙌🏻

**Review by GiorgiBeriashvili — APPROVED** [review]
Thank you for the PR!

I analyzed the code, I approve of the intention and think that it is a helpful feature. Though, I am unsure if `if (expected is MockDataCallback) return true;` is the right way to handle the matching since it will disregard any real matching logic and accept any `expected` argument that has a `MockDataCallback` method definition, but it can be handled later if problems arise.

Thus, I approve.

**Comment by GiorgiBeriashvili** [comment]
@theiskaa What do you think about hard-coded `return true` expression I mentioned [here](https://github.com/lomsa-dev/http-mock-adapter/pull/123#pullrequestreview-763509960)?

Do you have any suggestions/solutions or should we merge? I will leave it up to you. ^_^

**Comment by theiskaa** [comment]
Thank you @GiorgiBeriashvili. I don't think it would make any problem. So, we should merge for now, as you mentioned it can be handled later if problems arise.

**Pull request merged**